### PR TITLE
Ledger data export for non-determinism checks

### PR DESCRIPTION
**Pull request by miklos-da**
Summary of changes:
* Introduced `LedgerDataExporter` interface and its file-based implementation for storing submissions and resulting data committed to a ledger.
* Modified `BatchedSubmissionValidator` and friends to allow injecting an exporter instance.
* Added temporary `integrity-check-v2` build target that allows checking a ledger export file whether we are still generating the same output for a given input submission.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      println(s"${AnsiColor.GREEN}OK${AnsiColor.WHITE}")
+      true
+    } else {
+      println(s"${AnsiColor.RED}failed")
+      if (expectedWriteSet.size == actualWriteSet.size) {
+        for (((expectedKey, expectedValue), (actualKey, actualValue)) <- expectedWriteSet.zip(
+            actualWriteSet)) {
+          if (expectedKey == actualKey && expectedValue != actualValue) {
+            println(
+              s"expected value: ${bytesAsHexString(expectedValue)} vs. actual value: ${bytesAsHexString(actualValue)}")
```
Comment (mziolekda) [inline ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala:122]: Does scala string interpolator revert back to neutral color in a subsequent line after color has been switched? If not, I would suggest printing `ok` lines in neutral color and `nok` lines in red for faster analysis

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/Serialization.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.daml.ledger.participant.state
+import com.daml.ledger.participant.state.kvutils.export.FileBasedLedgerDataExporter.{
+  SubmissionInfo,
+  WriteSet
+}
+import com.google.protobuf.ByteString
+
+object Serialization {
+  def serializeSubmissionInfo(submissionInfo: SubmissionInfo, out: DataOutputStream): Unit = {
```
[the remark this answers is not in the document]
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/Serialization.scala]: Yes, that may be a path forward. I'm open to changing this wire format later -- the current design allows switching over to something else by simply changing the methods in this class only. For simplicity I chose DataInputStream/DataOutputStream as an initial implementation.

File: ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala (hunk 1 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala:99]: Unfortunately it doesn't revert back to neutral color, that's why I added a `println` to switch back to white before exiting. All `nok` lines are red and only `ok` is in green currently, i.e., the output looks like this:
image

Comment (mziolekda) [reply ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityChecker.scala:99]: That's ok, I did not read carefully enough. You only printing discrepancies to begin with.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/LogAppendingCommitStrategy.scala [hunk 3]
```diff
@@ -32,15 +34,21 @@ class LogAppendingCommitStrategy[Index](
         case (key, value) =>
           (keySerializationStrategy.serializeStateKey(key), Envelope.enclose(value))
       }(breakOut))
+      _ <- Future.successful(
+        ledgerDataExporter.addToWriteSet(correlationId, serializedKeyValuePairs))
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/LogAppendingCommitStrategy.scala:38]: Why not just use `_ =` here and below?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/Serialization.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/Serialization.scala]: Agreed, but I also think it's fine to follow up later.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/LogAppendingCommitStrategy.scala (hunk 3 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/LogAppendingCommitStrategy.scala]: Changed accordingly.

**Pull request merged**